### DISCUSSION TITLE
Partial `TypedDict` (`total=False`) and exception-based guard

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#7505] [Q&A]
The following issue: https://github.com/microsoft/pyright/issues/1739 was fixed by implementing exception-based guards. Would it make sense to restrict this guard to `KeyError` for `TypedDict`?

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjALABQDVANgIYDO7UAwgK7sxIAUQAeVABSiqxBDCRgUASgBcDKOqgIO7BkzacoAZWIw56duKIIylGgBpYYGK2YBeAGIv2xFWo0A3F15iZUwUOkZ6dlDjU1Q0LlcoAG8AX116GBA4VXoNKHYAbQAiQOZg4oBddQBiKBQwKGIQcBAHDiaRaxoyBmIxGXwpQfkUUKg6gHkQKABpYjghFtw-dS1ODKyc1YKSsorqifrG5ta%2BgdkefkEIKVz89Z16IA)

```python
from typing import TypedDict

class CustomExc(Exception):
    pass

class Settings(TypedDict, total=False):
    value: int

s: Settings = {}

try:
    s["value"]  # no error, as expected
except Exception:  # Or KeyError
    pass

try:
    s["value"]  # no error, but `CustomExc` does not inherit from `KeyError`
    # do other stuff, that might raise `CustomExc`
except CustomExc:
    pass
```

I'm also wondering if this feature should be documented [here](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards)?

Thanks in advance

**Comment by erictraut** [comment]
Pyright has no knowledge of specific exception classes, and adding such knowledge would be a slippery slope that I don't want to go down, so it's unlikely that we'd implement this suggestion.

If you want your code to be robust, I recommend against using try/except statements for normal code flow. They should be used only for exceptional cases. The preferred approach here is to use a normal `if` guard before accessing the non-required key in the TypedDict or accessing it with a `get` method that returns a reasonable default value.

```python
if 'value' in s:
    s['value']

s.get('value', 0)
```

**Reply by Viicos** [reply]
> Pyright has no knowledge of specific exception classes, and adding such knowledge would be a slippery slope that I don't want to go down, so it's unlikely that we'd implement this suggestion.

this makes sense, it is actually in my opinion a drawback of using `try..catch` as a guard for checking dict keys. In my case, I wasn't making use of such a feature, but I had the following code:

```python
try:
    val = do_something_that_might_raise_custom_exc()
    return some_func(val, settings["value"])
except CustomExc:
    ... 
```

Leading to a false negative on the `return` statement, as `settings["value"]` could raise a `KeyError` not handled by `except CustomExc` but `pyright` did not emit a diagnostic. Hence my feature request to only use the guard for `KeyError`'s, to avoid this false negative.

I ended up using the following anyway which is probably better:

```python
try:
    val = do_something_that_might_raise_custom_exc()
except CustomExc:
    ...
else:
    return some_func(val, settings["value"])  # correctly flagged by pyright: "value" might not exist on "settings"
```
